File: fsharp/grade-school/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School = Map.empty.Add(1, List.empty).Add(2, List.empty).Add(3, List.empty).Add(4, List.empty).Add(5, List.empty).Add(6, List.empty).Add(7, List.empty)

let add (student: string) (grade: int) (school: School): School =
    let sorted_grade_list = List.sort (student :: school.[grade])
    school.Add(grade, sorted_grade_list)
    

let roster (school: School): string list =  school.[1] @ school.[2] @ school.[3] @ school.[4] @ school.[5] @ school.[6] @ school.[7]

let grade (number: int) (school: School): string list = school.[number]

File: csharp/grade-school/GradeSchool.cs
using System;
using System.Collections.Generic;

public class GradeSchool
{
    private SortedDictionary<int, SortedSet<string>> roster = new SortedDictionary<int, SortedSet<string>>();
    public void Add(string student, int grade)
    {
        if (!roster.ContainsKey(grade)) roster[grade] = new SortedSet<string> { student };
        else roster[grade].Add(student);
    }

    public IEnumerable<string> Roster()
    {
        List<string> sortedNames = new List<string>();
        foreach (int grade in roster.Keys)
        {
            foreach (string name in roster[grade]) sortedNames.Add(name);
        }
        return sortedNames;
    }

    public IEnumerable<string> Grade(int grade)
    {
        if (roster.ContainsKey(grade)) return roster[grade];
        else return new List<string>();
    }
}